hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/PutAllOperation.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.multimap.impl.operations;
+
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.map.impl.MapEntries;
+import com.hazelcast.memory.NativeOutOfMemoryError;
+import com.hazelcast.multimap.impl.MultiMapContainer;
+import com.hazelcast.multimap.impl.MultiMapDataSerializerHook;
+import com.hazelcast.multimap.impl.MultiMapRecord;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.spi.impl.operationservice.MutatingOperation;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class PutAllOperation extends AbstractMultiMapOperation implements MutatingOperation {
+    private transient int currentIndex;
+    private MapEntries mapEntries;
+    private long recordId;
+
+    public PutAllOperation() {
+        super();
+    }
+
+    public PutAllOperation(String name, MapEntries mapEntries) {
+        super(name);
+        this.mapEntries = mapEntries;
+        //NB: general structure copied from c.hz.map.impl.operation.PutAllOperation
+        //May benefit from refactoring or default interfaces
+    }
+
+    @Override
+    public final void run() {
+        try {
+            runInternal();
+        } catch (NativeOutOfMemoryError e) {
+            //TODO: implement rerun?
+            e.printStackTrace();
+        }
+    }
+
+    protected void runInternal() {
+        //FIXME: potentially bad idiom?
+        int size = mapEntries != null ? mapEntries.size() : 0;
+        while (currentIndex < size) {
+            Data dataKey = mapEntries.getKey(currentIndex);
+            Data value = mapEntries.getValue(currentIndex);
+            put(dataKey, value);
+            currentIndex++;
+        }
+        //FIXME: is this the best place to set response?
+        response = true;
+    }
+
+    protected void put(Data dataKey, Data dataValue) {
+        MultiMapContainer container = getOrCreateContainer();
+        Collection c = (Collection) toObject(dataValue);
+        Collection<MultiMapRecord> coll = container.getOrCreateMultiMapValue(dataKey).getCollection(false);
+        Consumer addObjToCollection = o -> {
+            recordId = container.nextId();
+            MultiMapRecord record = new MultiMapRecord(recordId, o);
+            coll.add(record);
+            getOrCreateContainer().update();
+            publishEvent(EntryEventType.ADDED, dataKey, o, null);
+        };
+
+        if (coll instanceof Set) {
+            c.parallelStream().distinct().forEach(addObjToCollection);
```

Comment:
The `Set` and `List` implementations used internally in the multimap container are not thread-safe. Additionally, using `parallelStream` here breaks the principle of a single thread accessing a partition's data. Instead, do a plain `for` loop (and avoid the cost of stream & consumer allocations as a bonus). This is the reason of spurious test failures you observed. One more thing, event should be only triggered when a value is actually added (so adding an already existing value to a `Set`-backed multimap should not trigger an event).